In [1]:
import pandas as pd
pd.set_option('display.max_column', 20)

In [2]:
df = pd.read_excel('https://storage.googleapis.com/dqlab-dataset/cth_churn_analysis_train.xlsx')

In [3]:
print(df.head())

  ID_Customer Jenis_kelamin  umur  membership_program using_reward  \
0   1005-CTMP     Perempuan    41                  36           No   
1   1007-STSJ     Laki-laki    27                  36          Yes   
2   1012-NCGH     Laki-laki    45                  36           No   
3   1014-WWBN     Perempuan    33                  12           No   
4   1024-HYUA     Perempuan    38                  12           No   

      pembayaran Subscribe_brochure  harga_per_bulan  jumlah_harga_langganan  \
0           Cash                 No            10000                  360000   
1  Bank Transfer              Email            10000                  360000   
2           Cash                 No            10000                  360000   
3  Bank Transfer                Yes            10000                  120000   
4           Cash                 No            10000                  120000   

  churn  
0   Yes  
1   Yes  
2    No  
3    No  
4    No  


In [4]:
#menghilangkan kolom 'ID_Customer' dari data frame dikarenakan kolom ini tidak relevan untuk dijadikan input dalam tugas klasifikasi (ID customer tidak mempengaruhi apakah customer akan lanjut berlangganan atau tidak
df.drop('ID_Customer', axis=1, inplace=True)
print(df)

    Jenis_kelamin  umur  membership_program using_reward     pembayaran  \
0       Perempuan    41                  36           No           Cash   
1       Laki-laki    27                  36          Yes  Bank Transfer   
2       Laki-laki    45                  36           No           Cash   
3       Perempuan    33                  12           No  Bank Transfer   
4       Perempuan    38                  12           No           Cash   
..            ...   ...                 ...          ...            ...   
494     Perempuan    42                  24           No           Cash   
495     Perempuan    27                  24          Yes  Bank Transfer   
496     Laki-laki    25                   6           No  Bank Transfer   
497     Laki-laki    35                   6           No    Credit Card   
498     Perempuan    45                  12          Yes           Cash   

    Subscribe_brochure  harga_per_bulan  jumlah_harga_langganan churn  
0                   No     

In [5]:
#menyimpan kolom 'churn' sebagai list ke dalam variabel y
y = df.pop('churn').to_list()

#mengubah nilai 'Yes' menjadi 1 dan nilai 'No' menjadi 0 agar sesuai dengan format yang sebelumnya telah kita bahas
y = [1 if label == 'Yes' else 0 for label in y]
print(y)

[1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1, 

In [6]:
#memeriksa tipe data dari setiap kolom
print('Tipe data setiap kolom:')
print('-----------------------')
df.info()

#lakukan pengecekan untuk kolom dengan tipe data 'object' (kategorikal)
print('\nKolom dengan tipe data object (kategorikal):')
print('--------------------------------------------')
for col in df.select_dtypes(include=['object']):
	print(df[col].value_counts())
	print("===============================")

#statistik deskriptif dari setiap kolom
print('\nStatistik deskriptif dari setiap kolom:')
print('---------------------------------------')
print(df.describe())

Tipe data setiap kolom:
-----------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 499 entries, 0 to 498
Data columns (total 8 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Jenis_kelamin           499 non-null    object
 1   umur                    499 non-null    int64 
 2   membership_program      499 non-null    int64 
 3   using_reward            499 non-null    object
 4   pembayaran              499 non-null    object
 5   Subscribe_brochure      499 non-null    object
 6   harga_per_bulan         499 non-null    int64 
 7   jumlah_harga_langganan  499 non-null    int64 
dtypes: int64(4), object(4)
memory usage: 31.3+ KB

Kolom dengan tipe data object (kategorikal):
--------------------------------------------
Laki-laki    263
Perempuan    236
Name: Jenis_kelamin, dtype: int64
No     261
Yes    238
Name: using_reward, dtype: int64
Bank Transfer    195
Cash             156
Credit Card      148


In [7]:
#membuang kolom 'harga_per_bulan'
df.drop('harga_per_bulan', axis=1, inplace=True)

#membuang kolom 'jumlah_harga_langganan'
df.drop('jumlah_harga_langganan', axis=1, inplace=True)
df.info()

#note: 
#jika suatu atribut hanya terdiri atas satu nilai saja, 
#maka atribut ini dapat dianggap tidak berguna dikarenakan saat mesin melakukan proses pembelajaran, 
#mesin akan mencari ciri (perbedaan nilai yang dimiliki oleh setiap label). 
#Saat  nilai dari suatu atribut hanya terdiri dari satu nilai saja, 
#mesin tidak dapat mempelajari hal apapun berdasarkan ciri tersebut.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 499 entries, 0 to 498
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Jenis_kelamin       499 non-null    object
 1   umur                499 non-null    int64 
 2   membership_program  499 non-null    int64 
 3   using_reward        499 non-null    object
 4   pembayaran          499 non-null    object
 5   Subscribe_brochure  499 non-null    object
dtypes: int64(2), object(4)
memory usage: 23.5+ KB


In [8]:
#mengimport class LabelEncoder untuk mengubah atribut dengan dua kemungkinan nilai (binary)
from sklearn.preprocessing import LabelEncoder
 
#menyiapkan dictionary untuk menyimpan seluruh LabelEncoder untuk setiap atribut kategorikal yang bersifat biner
labelers = {}
 
#untuk setiap kolom dengan tipe data 'object' (kategorikal)
column_categorical_non_binary = []
for col in df.select_dtypes(include=['object']):
	#saat jumlah nilai unik dari suatu kolom sama dengan dua
	#atau dengan kata lain kolom bersifat biner
	if len(df[col].unique()) == 2:
		#buat objek LabelEncoder baru untuk kolom dan tampung dalam
		#dictionary labelers
		labelers[col] = LabelEncoder()
		#meminta objek LabelEncoder untuk mempelajari dan
		#mentransformasikan kolom
		df[col] = labelers[col].fit_transform(df[col])
	#untuk kolom bersifat non-biner
	else:
		#tambahkan nama kolom ke dalam array yang telah disiapkan
		column_categorical_non_binary.append(col)		
print(df.head())

df = pd.get_dummies(df, columns= column_categorical_non_binary)
print(df.head())

   Jenis_kelamin  umur  membership_program  using_reward     pembayaran  \
0              1    41                  36             0           Cash   
1              0    27                  36             1  Bank Transfer   
2              0    45                  36             0           Cash   
3              1    33                  12             0  Bank Transfer   
4              1    38                  12             0           Cash   

  Subscribe_brochure  
0                 No  
1              Email  
2                 No  
3                Yes  
4                 No  
   Jenis_kelamin  umur  membership_program  using_reward  \
0              1    41                  36             0   
1              0    27                  36             1   
2              0    45                  36             0   
3              1    33                  12             0   
4              1    38                  12             0   

   pembayaran_Bank Transfer  pembayaran_Cash  pemb

In [9]:
#merepresentasikan seluruh kolom dalam df sebagai matrik dan menampungnya dalam variabel X.
X = df.to_numpy()
 
#menampilkan informasi dimensi dari variabel X
print("Dimensi dari variabel X:",X.shape)

Dimensi dari variabel X: (499, 10)


In [10]:
#library yang akan kita gunakan untuk membagi dataset
from sklearn.model_selection import train_test_split
 
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.1, random_state=23)
 
print("Dimensi data X mula-mula:", X.shape)
print("Dimensi data y mula-mula:",len(y),"\n")
 
print("Dimensi data X train:", X_train.shape)
print("Dimensi data y train:", len(y_train),"\n")
 
print("Dimensi data X test:", X_test.shape)
print("Dimensi data y test:", len(y_test),"\n") 

Dimensi data X mula-mula: (499, 10)
Dimensi data y mula-mula: 499 

Dimensi data X train: (449, 10)
Dimensi data y train: 449 

Dimensi data X test: (50, 10)
Dimensi data y test: 50 



In [11]:
#untuk melakukan proses klasifikasi menggunakan algoritma DecisionTree kita dapat mengimport object 'DecisionTreeClassifier' pada modul 'tree' milik library scikit-learn
from sklearn.tree import DecisionTreeClassifier
 
#potongan kode ini dapat kita gunakan untuk menginisialisasi model dengan default parameter yang telah disediakan oleh library perhatikan bahwa parameter random_state hanya digunakan untuk reproducibility seperti layaknya pada fungsi train_test_split
model = DecisionTreeClassifier(random_state=57)
 
#meminta model untuk melatih dirinya menggunakan data X_train dan y_train yang telah disiapkan
print("Memulai melatih 'model'.")
model.fit(X_train, y_train)
print("Selesai melatih 'model'.")

Memulai melatih 'model'.
Selesai melatih 'model'.


In [12]:
X_new = [
	[1, 36, 12, 0, 0, 1, 0, 0, 0, 1], # <= data pertama
	[0, 23, 24, 1, 1, 0, 0, 0, 1, 0]  # <= data kedua
]

#melakukan prediksi pada data dalam variabel X_new
y_new_pred = model.predict(X_new)
print(y_new_pred)

[1 0]


Output di atas menandakan bahwa tidak terdapat potensi berhenti berlangganan untuk data kedua dan terdapat potensi berlangganan pada data pertama.

In [13]:
#meng-import fungsi untuk menghitung akurasi pada library scikit-learn
from sklearn.metrics import accuracy_score

#melakukan prediksi terhadap data latih dan menampung
#label hasil prediksi dalam variabel y_pred
y_pred = model.predict(X_train)
score = accuracy_score(y_train, y_pred)
print("Skor akurasi untuk data latih:", score)

#melakukan prediksi terhadap data testing dan menampung
#label hasil prediksi dalam variabel y_pred
y_pred = model.predict(X_test)
score = accuracy_score(y_test,y_pred)
print("Skor akurasi untuk data testing:", score)

Skor akurasi untuk data latih: 0.9710467706013363
Skor akurasi untuk data testing: 0.4


Model klasifikasi yang dikembangkan memiliki memiliki akurasi sebesar 97.10%. Sedangkan data testing model memiliki akurasi 40%.
Performa dari model yang telah dilatih akan dibandingkan dengan dengan model cap-cip-cup atau model yang menebak label dari sebuah data yang diberikan secara sembarangan (random guessing).

Daripada menggunakan model klasifikasi, status keberlanjutan berlangganan ditentukan secara asal. Untuk proses klasifikasi biner, jika satu dari dua label dipilih secara acak, maka peluang prediksi akan benar sama dengan satu label yang dipilih dibandingkan dengan dua pilihan yang ada (P(true)=1/2 atau 50%)

Dari proses data dalam memprediksi data testing di atas, model yang dibuat bahkan memprediksi 10% lebih buruk dibandingkan dengan proses cap-cip-cup! Tapi untuk data latih, model dapat dibilang baik karena model memprediksi 47.10% lebih baik

Dalam kasus ini dimana mendapat skor model yang sangat baik untuk data training namun buruk untuk data testing dikarenakan performa pada data testing mencerminkan performa model untuk data-data yang belum pernah ia kenali. Kemudian, saat memanfaatkan model untuk melakukan proses klasifikasi pada data baru, sangat kecil kemungkinan data tersebut tercermin dalam data latih.

In [14]:
from sklearn.metrics import accuracy_score

#mendefinisikan nilai dari parameter 'min_samples_split' yang akan dicobakan
min_samples_split_search = [2,4,8,16,32,64]

#mendefinisikan nilai dari parameter 'max_depth' yang akan dicobakan
max_depth_search = [4,8,16,32,64,128]

#mendefinisikan variabel untuk menyimpan skor terbaik dari setiap model dengan parameter yang berbeda
max_score = 0

#mendefinisikan variabel untuk menyimpan model terbaik.
best_model = None

#mencoba membuat model DecisionTree berdasarkan nilai kombinasi
#dari parameter 'min_samples_split' dan 'max_depth'
for ms in min_samples_split_search:
	for md in max_depth_search:
		# menginisialisasi model berdasarkan salah satu kombinasi nilai yang ada
		model = DecisionTreeClassifier(min_samples_split=ms, max_depth=md, random_state=57)
		
		#melatih model yang telah diinisialisasi dengan data 
		#X_train dan label y_train
		model.fit(X_train,y_train)
		
		#melakukan prediksi terhadap data X_test
		y_pred = model.predict(X_test)
		
		#menghitung skor berdasarkan nilai aktual (y_test) dan (y_pred)
		score = accuracy_score(y_test, y_pred)
		
		#jika score yang dihasilkan oleh model lebih besar dari skor
		#terbesar yang dicatat (max_score), maka
		if max_score < score:
			#simpan model dalam variabel best_model
			best_model = model
			
			#update max_score menjadi score milik model
			max_score = score
			
print("Skor testing terbaik: ", max_score)
print("Parameter model: max_depth=", 
      best_model.get_params()['max_depth'], 
      ", min_samples_split=",
      best_model.get_params()['min_samples_split'])

Skor testing terbaik:  0.58
Parameter model: max_depth= 8 , min_samples_split= 32


Mengetahui bahwa skor pada data testing lebih penting dibanding skor pada data training. Maka dari itu, ia menggunakan variabel “max_score” untuk menyimpan akurasi terbaik untuk data testing seperti yang dapat terlihat pada potongan kode di atas. 

Melalui proses pencarian parameter terbaik untuk model Decision Tree,berhasil meningkatkan akurasi sampai dengan 18% (dari 40% sampai dengan 58%). Model Decision Tree terbaik dihasilkan saat nilai parameter max_depth sama dengan delapan (8) dan min_samples_split sama dengan 32.